src/memory.jl

```diff
@@ -176,6 +176,26 @@ function free(buf::HostBuffer)
     end
 end
 
+##
+# Pinned memory management
+#
+# - IdDict does not free the memory
+# - WeakRef dict does not unique the key by objectid
+# TODO:
+# - Deal with memory regions/views
+const __pinned_memory = Dict{Ptr, WeakRef}()
+
+function pin(a::Array, flags=0)
+    # use pointer instead of objectid?
+    ptr = pointer(a)
+    if haskey(__pinned_memory, ptr) && __pinned_memory[ptr].value !== nothing
+        return nothing
```

Comment:
```suggestion
        @assert __pinned_memory[ptr].value === a
        return nothing
```

I don't really get this though, so you're not allowed to register memory twice, but wat if a Julia array is freed (`.value === nothing`) but the memory reused; won't you then end up registering the same pointer, now backed by a different array, twice?

Reply:
Right, Looking at UCX they have an `rcache` which is a cache for memory regions that are registered, and that supports merging of regions.
I was hoping to come up with a less complex implementation, but I am not a 100% how to best deal with `reallocs`, and free's and allocs that are faster than the finalizers. Since @kshyatt reported that the `Set` implementation can cause finalizer errors I suspect we need to be more careful, since it basically means that we can see GC freeing an array an then re-using its memory before we get to run the finalizer.

Reply:
FWIW using this code "in production" with many thousands of really big arrays (I can try to check my memory footprint) I haven't hit this issue.